doc/in-progress/object_versions.md

```diff
@@ -0,0 +1,147 @@
+
+## Move Objects
+
+Move Object versions are calculated using a variant of [Lamport timestamps](https://en.wikipedia.org/wiki/Lamport_timestamp) that guarantees that versions never get re-used in all cases:  The new version for objects touched by a transaction is one greater than the max version among all input objects to the transaction.  For example, a transaction transfering an object `O` at version `5` using a gas object `G` at version `3` updates both `O` and `G`'s version to `1 + max(5, 3) = 6`.
+
+The relevance of versions for accessing an object as a transaction input changes depending on that object's ownership:
+
+### Address-owned Objects
+
+Address-owned transaction inputs must be referenced at a specific ID and version.  When a validator signs a transaction with an owned object input at a specific version, that version of the object is **locked** to that transaction, from the validator's perspective: it rejects requests to sign other transactions that require the same input (same ID and version).
+
+If `F + 1` validators sign one transaction that takes an object as input, and a different `F + 1` validators sign a different transaction that takes the same object as input, that object (and all the other inputs to both transactions) are **equivocated** meaning they cannot be used for any further transactions in that epoch.  This is because neither transaction can form a quorum without relying on a signature from a validator that has already committed the object to a different transaction, which it cannot get.  All locks are reset at the end of the epoch, which frees the objects up once more.
+
+Only an object's owner can equivocate it, but this is not a desirable thing to do.  Equivocation can be avoided by carefully managing the versions of address-owned input objects.
+
+### Immutable Objects
+
+Like Address-owned objects, immutable objects are also referenced at an ID and version, but they do not need to be locked, because their contents and versions do not change.  Their version is relevant because they could have started life as an Address-owned object and later been frozen. The given version identifies the point at which they became immutable.
+
+### Shared Objects
+
+Specifying a shared transaction input is slightly more complex.  It is referenced by its ID, the version it was shared at, and a flag indicating whether it is accessed mutably.  The precise version the transaction will access is **not** specified, because it is decided by consensus during transaction scheduling: When scheduling multiple transactions that touch the same shared object, validators agree the order of those transactions, and pick each transaction's input versions for the shared object's accordingly (one transaction's output version becomes the next transaction's input version, and so on).
+
+Shared transaction inputs that are referenced immutably participate in scheduling, but don't modify the object or increment its version.
+
+### Wrapped Objects
```

Comment:
@tnowacki / @gdanezis , curious for your thoughts on wrapped objects and dynamic fields.